### ISSUE TITLE
`__STRICT_ANSI__' not defined

### ISSUE THREAD

**Issue by martinergb** [sakra/cotire#94]
I got this warning when compiling on Ubuntu and the whole project is built without pch as normal build:

`warning: /home/utility/cotire/IO.Utility_CXX_prefix.hxx.gch: not used because '__STRICT_ANSI__' not defined [-Winvalid-pch]`

So is there a way to fix this warning?

I don't have any issues when compiling on Windows.

**Comment by sakra** [comment]
Do you have `__STRICT_ANSI__` as an `add_definition` in your CMakeLists.txt?

**Comment by martinergb** [comment]
No, I don't have this definition in my CMakeLists.txt

I try to add this definition by:

`add_definitions(-D__STRICT_ANSI__)`

I have new warning:

`warning: /home/utility/cotire/IO.Utility_CXX_prefix.hxx.gch: not used because 'linux' is defined [-Winvalid-pch]`

**Comment by sakra** [comment]
Please double check that you do not modify any target properties that affect compilation after you have applied cotire.

**Comment by martinergb** [comment]
I found out that this issue caused by

`set_property(TARGET example PROPERTY CXX_STANDARD 11)`
`set_property(TARGET example PROPERTY CXX_STANDARD_REQUIRED ON)`

But I cannot remove these set commands because I use lots of c++11 features in my project. What can I do now? Thank you.

**Comment by sakra** [comment]
Please gist the output of a `cmake -DCOTIRE_DEBUG=On` run.

**Comment by martinergb** [comment]
Below is the output with debug on:

admin@ubuntu:~/Documents/cotire-master/build$ cmake ../ -DCOTIRE_DEBUG=On
-- The C compiler identification is GNU 4.8.4
-- The CXX compiler identification is GNU 4.8.4
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Boost version: 1.54.0
-- Found the following Boost libraries:
--   date_time
--   filesystem
--   system
--   regex
/usr/include
-- cotire 1.7.8 loaded.
-- C source file extensions: c;m
-- C ignore extensions: h;H;o;O;obj;OBJ;def;DEF;rc;RC
-- C exclude extensions: m;mm
-- CXX source file extensions: C;M;c++;cc;cpp;cxx;mm;CPP
-- CXX ignore extensions: inl;h;hpp;HPP;H;o;O;obj;OBJ;def;DEF;rc;RC
-- CXX exclude extensions: m;mm
-- Filtered example CXX sources: main.cpp;example.cpp;log.cpp
-- CXX target example cotired.
-- CXX source file extensions: C;M;c++;cc;cpp;cxx;mm;CPP
-- CXX ignore extensions: inl;h;hpp;HPP;H;o;O;obj;OBJ;def;DEF;rc;RC
-- CXX exclude extensions: m;mm
-- Filtered example CXX sources: main.cpp;example.cpp;log.cpp
-- # cotire.cmake 1.7.8 generated file
# /home/admin/Documents/cotire-master/build/src/example_CXX_cotire.cmake
set (COTIRE_ADDITIONAL_PREFIX_HEADER_IGNORE_EXTENSIONS "inc;inl;ipp")
set (COTIRE_ADDITIONAL_PREFIX_HEADER_IGNORE_PATH "")
set (COTIRE_CLEAN_ALL_TARGET_NAME "clean_cotire")
set (COTIRE_CLEAN_TARGET_SUFFIX "_clean_cotire")
set (COTIRE_CMAKE_MODULE_FILE "/home/admin/Documents/cotire-master/CMake/cotire.cmake")
set (COTIRE_CMAKE_MODULE_VERSION "1.7.8")
set (COTIRE_DEBUG "On")
set (COTIRE_INTDIR "cotire")
set (COTIRE_MAXIMUM_NUMBER_OF_UNITY_INCLUDES "0")
set (COTIRE_MINIMUM_NUMBER_OF_TARGET_SOURCES "3")
set (COTIRE_PCH_ALL_TARGET_NAME "all_pch")
set (COTIRE_PCH_TARGET_SUFFIX "_pch")
set (COTIRE_PREFIX_HEADER_FILENAME_SUFFIX "_prefix")
set (COTIRE_TARGETS_FOLDER "cotire")
set (COTIRE_TARGET_CONFIGURATION_TYPES "None")
set (COTIRE_TARGET_CXX_COMPILER_LAUNCHER "COTIRE_TARGET_CXX_COMPILER_LAUNCHER-NOTFOUND")
set (COTIRE_TARGET_IGNORE_PATH "/home/admin/Documents/cotire-master")
set (COTIRE_TARGET_INCLUDE_PRIORITY_PATH "")
set (COTIRE_TARGET_LANGUAGE "CXX")
set (COTIRE_TARGET_MAXIMUM_NUMBER_OF_INCLUDES "")
set (COTIRE_TARGET_POST_UNDEFS "")
set (COTIRE_TARGET_PRE_UNDEFS "")
set (COTIRE_TARGET_SOURCES "main.cpp;example.cpp;log.cpp")
set (COTIRE_UNITY_BUILD_ALL_TARGET_NAME "all_unity")
set (COTIRE_UNITY_BUILD_TARGET_SUFFIX "_unity")
set (COTIRE_UNITY_OUTPUT_DIRECTORY "")
set (COTIRE_UNITY_SOURCE_EXCLUDE_EXTENSIONS "m;mm")
set (COTIRE_UNITY_SOURCE_FILENAME_SUFFIX "_unity")
set (CMAKE_GENERATOR "Unix Makefiles")
set (CMAKE_BUILD_TYPE "")
set (CMAKE_CXX_COMPILER_ID "GNU")
set (CMAKE_CXX_COMPILER_VERSION "4.8.4")
set (CMAKE_CXX_COMPILER "/usr/bin/c++")
set (CMAKE_CXX_COMPILER_ARG1 "")
set (CMAKE_INCLUDE_FLAG_CXX "-I")
set (CMAKE_INCLUDE_SYSTEM_FLAG_CXX "-isystem ")
set (CMAKE_CXX_SOURCE_FILE_EXTENSIONS "C;M;c++;cc;cpp;cxx;mm;CPP")

-- example unity source max includes: 0
-- unity files: /home/admin/Documents/cotire-master/build/src/cotire/example_CXX_unity.cxx
-- add_custom_command: OUTPUT /home/admin/Documents/cotire-master/build/src/cotire/example_CXX_unity.cxx COMMAND /usr/local/bin/cmake;--warn-uninitialized;-DCOTIRE_BUILD_TYPE:STRING=$<CONFIGURATION>;-DCOTIRE_VERBOSE:BOOL=$(VERBOSE);-P;/home/admin/Documents/cotire-master/CMake/cotire.cmake;unity;/home/admin/Documents/cotire-master/build/src/example_CXX_cotire.cmake;/home/admin/Documents/cotire-master/build/src/cotire/example_CXX_unity.cxx DEPENDS /home/admin/Documents/cotire-master/build/src/example_CXX_cotire.cmake
-- add_custom_command: OUTPUT /home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.cxx COMMAND /usr/local/bin/cmake;--warn-uninitialized;-DCOTIRE_BUILD_TYPE:STRING=$<CONFIGURATION>;-DCOTIRE_VERBOSE:BOOL=$(VERBOSE);-P;/home/admin/Documents/cotire-master/CMake/cotire.cmake;prefix;/home/admin/Documents/cotire-master/build/src/example_CXX_cotire.cmake;/home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.cxx;/home/admin/Documents/cotire-master/build/src/cotire/example_CXX_unity.cxx DEPENDS /home/admin/Documents/cotire-master/build/src/cotire/example_CXX_unity.cxx 
-- add_custom_command: OUTPUT /home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.hxx COMMAND /usr/local/bin/cmake;--warn-uninitialized;-DCOTIRE_BUILD_TYPE:STRING=$<CONFIGURATION>;-DCOTIRE_VERBOSE:BOOL=$(VERBOSE);-P;/home/admin/Documents/cotire-master/CMake/cotire.cmake;combine;/home/admin/Documents/cotire-master/build/src/example_CXX_cotire.cmake;/home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.hxx;/home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.cxx DEPENDS /home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.cxx
-- add_custom_command: OUTPUT /home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.hxx.gch /usr/local/bin/cmake;--warn-uninitialized;-DCOTIRE_BUILD_TYPE:STRING=$<CONFIGURATION>;-DCOTIRE_VERBOSE:BOOL=$(VERBOSE);-P;/home/admin/Documents/cotire-master/CMake/cotire.cmake;precompile;/home/admin/Documents/cotire-master/build/src/example_CXX_cotire.cmake;/home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.hxx;/home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.hxx.gch;main.cpp DEPENDS /home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.hxx IMPLICIT_DEPENDS CXX /home/admin/Documents/cotire-master/build/src/cotire/example_CXX_prefix.hxx
-- CXX source file extensions: C;M;c++;cc;cpp;cxx;mm;CPP
-- CXX ignore extensions: inl;h;hpp;HPP;H;o;O;obj;OBJ;def;DEF;rc;RC
-- CXX exclude extensions: m;mm
-- Filtered example CXX sources: main.cpp;example.cpp;log.cpp
-- add target EXECUTABLE example_unity  EXCLUDE_FROM_ALL log.h;example.h;/home/admin/Documents/cotire-master/build/src/cotire/example_CXX_unity.cxx
-- unity target example_unity link strategy: COPY_UNITY
-- Configuring done
-- Generating done
-- Build files have been written to: /home/admin/Documents/cotire-master/build

**Comment by sakra** [comment]
Please also post your modified CMakeLists.txt from the example.

**Comment by martinergb** [comment]
I removed all message() commands for clarity, so there are just few lines:

`add_executable(example main.cpp example.cpp log.cpp log.h example.h)`

`cotire(example)`
`set_property(TARGET example PROPERTY CXX_STANDARD 11)`
`set_property(TARGET example PROPERTY CXX_STANDARD_REQUIRED ON)`

**Comment by sakra** [comment]
And here is your problem. Cotire must be applied *after* the properties `CXX_STANDARD` and `CXX_STANDARD_REQUIRED` are set, i.e.:

```
set_property(TARGET example PROPERTY CXX_STANDARD 11)
set_property(TARGET example PROPERTY CXX_STANDARD_REQUIRED ON)
cotire(example)
```

Also see the [manual section on common pitfalls](https://github.com/sakra/cotire/blob/master/MANUAL.md#do-not-modify-a-targets-build-related-properties-after-applying-cotire). Under Windows it works by chance because these properties do not affect compilation when using Visual Studio.

**Comment by martinergb** [comment]
Thank you, I try to set those properties before applying cotire but I sill get the same issue

**Comment by sakra** [comment]
Try recreating your build folder.

**Comment by harningt** [comment]
Have run into this, including with clean build - slightly different in that I set the CXX_STANDARD / CXX_STANDARD_REQUIRED flags globally.

**Comment by krey** [comment]
+1

After adding the definition
`add_definitions(-D__STRICT_ANSI__)`
my error changed to
`not used because '__GLIBCXX_BITSIZE_INT_N_0' is defined [-Winvalid-pch]`

The workaround for me was commenting out the CXX_STANDARD_REQUIRED line:
```
set (CMAKE_CXX_STANDARD 14)
#set (CMAKE_CXX_STANDARD_REQUIRED ON)
```

**Comment by sakra** [comment]
Fixed in 1.7.9.
